kernel/src/util/iovec.rs

```diff
@@ -37,92 +39,190 @@ impl TryFrom<UserIoVec> for IoVec {
+                .read_val()?;
+            IoVec::try_from(uiov)?
+        };
+
+        if iov.is_empty() {
+            continue;
+        }
+
+        let converted = convert_iovec(&iov, vm_space)?;
+        v.push(converted)
     }
 
-    /// Reads bytes from the user space buffer pointed by
-    /// the `IoVec` to `dst`.
-    ///
-    /// If successful, the read length will be equal to `dst.len()`.
-    ///
-    /// # Panics
-    ///
-    /// This method will panic if
-    /// 1.`dst.len()` is not the same as `self.len()`;
-    /// 2. `self.is_empty()` is `true`.
-    pub fn read_exact_from_user(&self, dst: &mut [u8]) -> Result<()> {
-        assert_eq!(dst.len(), self.len);
-        assert!(!self.is_empty());
+    Ok(v.into_boxed_slice())
+}
+
+/// A collection of [`VmReader`]s.
+///
+/// Such readers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecReader<'a>(Box<[VmReader<'a>]>);
 
-        CurrentUserSpace::get().read_bytes(self.base, &mut VmWriter::from(dst))
+/// A collection of [`VmWriter`]s.
+///
+/// Such writers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecWriter<'a>(Box<[VmWriter<'a>]>);
+
+impl<'a> IoVecReader<'a> {
+    /// Creates a new `IoVecReader` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let readers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::reader)?;
+        Ok(Self(readers))
     }
 
-    /// Writes bytes from the `src` buffer
-    /// to the user space buffer pointed by the `IoVec`.
-    ///
-    /// If successful, the written length will be equal to `src.len()`.
-    ///
-    /// # Panics
-    ///
-    /// This method will panic if
-    /// 1. `src.len()` is not the same as `self.len()`;
-    /// 2. `self.is_empty()` is `true`.
-    pub fn write_exact_to_user(&self, src: &[u8]) -> Result<()> {
-        assert_eq!(src.len(), self.len);
-        assert!(!self.is_empty());
+    pub fn readers_mut(&'a mut self) -> &'a mut [VmReader<'a>] {
+        &mut self.0
+    }
+}
 
-        CurrentUserSpace::get().write_bytes(self.base, &mut VmReader::from(src))
+impl<'a> IoVecWriter<'a> {
+    /// Creates a new `IoVecWriter` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let writers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::writer)?;
+        Ok(Self(writers))
     }
 
-    /// Reads bytes to the `dst` buffer
-    /// from the user space buffer pointed by the `IoVec`.
-    ///
-    /// If successful, returns the length of actually read bytes.
-    pub fn read_from_user(&self, dst: &mut [u8]) -> Result<usize> {
-        let len = self.len.min(dst.len());
-        CurrentUserSpace::get().read_bytes(self.base, &mut VmWriter::from(&mut dst[..len]))?;
-        Ok(len)
+    pub fn writers_mut(&'a mut self) -> &'a mut [VmWriter<'a>] {
```

Comment:
Look at the following user code.

```
let writer = IoVecWriter::copy_from_user(ctx, start_addr, count)?;
let writers = writer.writers_mut();
```

The code looks confusing because in the two consecutive lines, you get two different kinds of writer. This is very confusing. Yes, it is used to represent `io_vec` in C. But we don't need to call it `IoVecXxx`. I think calling it `VmWriterArray` is more appropriate. (It is not a vector because it cannot adjust sizes).

```rust
let writer_array = VmWriterArray::copy_from_user(ctx, start_addr, count)?;
let writers = writer_array.writers_mut();
```

You see. Now the user code looks much natural. You have an array first. Then you get elements from the array. No surprises.

Reply:
Agree with the point.

> But we don't need to call it IoVecXxx.

I think this will also cause confusion. How about `VmWriterArray::from_user_io_vecs`?

Reply:
The two renaming are both adopted. 

`IoVecWriter` -> `VmWriterArray`
`VmWriterArray::copy_from_user` -> `VmWriterArray::from_user_io_vecs`

I also renamed the trait `IoVecWrite` as `MultiWrite`, since `IoVec` prefix is not used now.